src/Pact/Native.hs

```diff
@@ -1445,3 +1454,91 @@ continueNested i as = gasUnreduced i as $ case as of
   unTVar = \case
     TVar (Ref d) _ -> unTVar d
     d -> d
+
+-- | A tag for determining how to proceed when encountering an invalid
+-- base64-encoded message.
+--
+-- Legacy:
+--   Although we are using base64-bytestring > 1.0, emulate the behavior of
+--   base64-bytestring-0.1, for hash compatibility with historical blocks.
+--   Messages that continue to fail with new error messages will have those
+--   error messages parsed and reformatted into the older form. And messages
+--   that now fail due to "non-canonical encoding" will be parsed again
+--   leniently, because the legacy base64 parser accepted these messages.
+--
+--  Simplified:
+--    Only base64-encoded messages that pass strict parsing will be accepted
+--    (no second lenient pass for non-canonical encodings). All failures to
+--    parse will result in the same single error message. This makes the error
+--    messages less informative, but makes it easier to maintain compatibility
+--    as the base64 parsing algorithm evolves, for failures encountered after
+--    the fork that enables Simplified error messages..
+data Base64DecodingBehavior
+  = Legacy
+  | Simplified
+  deriving (Eq, Show)
+
+-- | Convert from base64-bytestring-1.0 behavior to
+-- base64-bytestring-0.1 behavior. This is needed in order
+-- to preserve hash equality with old versions of pact (which
+-- used base64-bytestring-0.1).
+--
+-- Throws a Pact `evalError` if it fails to parse an encountered
+-- base64-decoding error message.
+base64DecodeWithShimmedErrors
+  :: Info
+  -> Base64DecodingBehavior
+  -> Text
+  -> Eval e (Either String BS.ByteString)
+base64DecodeWithShimmedErrors i behavior txt =
+
+  -- Attempt to decode the bytestring, and convert error messages to Text.
+  case first Text.pack $ parseB64UrlUnpaddedText' txt of
+
+    -- base64-bytestring-0.1 is more strict than base64-bytestring-0.1,
+    -- so all new successful decodings succeeded on the old version, too.
+    Right e -> return $ Right e
+
+    -- With Simplified error messages, map every error to a single string.
+    Left _ | behavior == Simplified ->
+             return $ Left "Could not base64-decode string"
```

Comment:
weird indentations here and below on some things

Reply:
Better now?